readthedocs/doc_builder/python_environments.py

```diff
@@ -72,6 +74,103 @@ def delete_existing_venv_dir(self):
             )
             shutil.rmtree(venv_dir)
 
+    def install_languages(self):
```

Comment:
I might suggest we use something more specific than `languages` in our nomenclature for the feature here. We use `language`/`lang` already as a core part of our product. On it's own `languages` could mean a number of thing and will likely get confusing.

Reply:
Makes sense. I will use the same name as in the config file: `build_languages`.

Reply:
This seems better than "languages". I would still be interested in maybe finding another way to refer to all of this, but don't feel super strongly.

Are we only ever going to cache language runtimes with this? This feels more like a "dependency cache" or "precompile cache" or something similar. Other services use nomenclature like "build packs" for similar installable chunks.

If we definitely don't see our needs shifting much away from just language runtimes, would "runtimes" fit better? `build.runtimes` is maybe a little bit more clear to users than `build.languages`, it could help resolve some of the issues with overloading "languages" in our list of terms.

Reply:
> Are we only ever going to cache language runtimes with this? This feels more like a "dependency cache" or "precompile cache" or something similar. Other services use nomenclature like "build packs" for similar installable chunks.

My idea is to only cache languages installed by `asdf` here. Nothing else.

For cache other different things (build packs, etc) we have another storage bucket called `envs` that feels like a good fit for that.

I want to explicitly have one bucket for these `asdf` languages/binaries that is not mixed with anything else.

> If we definitely don't see our needs shifting much away from just language runtimes, would "runtimes" fit better? `build.runtimes` is maybe a little bit more clear to users than `build.languages`, it could help resolve some of the issues with overloading "languages" in our list of terms.

I personally don't see `build.runtimes` clearer than `build.languages`; but I don't have a strong opinion here. I understand the problem of using `build.languages` and I'm happy if we can find something that works better, but I don't have a good idea, tho 😢 --I'm fine using `build.runtimes` to avoid the confusion.

Reply:
> we have another storage bucket called envs

Well, that's a different level of caching, which we could repurpose as we decided not to use this bucket.

> I personally don't see build.runtimes clearer than build.languages; but I don't have a strong opinion here.

Same, I don't have any strong feelings about this either, it was just a thought. I don't yet have any better opinions here. I'm mostly +0 on `build.languages`, but want to take some extra time here getting this right the first time, as changing our config file in the future is always going to be painful.

Reply:
> Same, I don't have any strong feelings about this either, it was just a thought. I don't yet have any better opinions here. I'm mostly +0 on `build.languages`, but want to take some extra time here getting this right the first time, as changing our config file in the future is always going to be painful.

I agree. Pinging @astrojuanlu @ericholscher @stsewd for some input here.

Reply:
I also felt that languages wasn't clear enough, as we basically are only installing packages (`build.packages?`), nothing special around that.

In our code we use "environment" to refer to these (python env, conda env). But that can be confused with env vars.

Reply:
could also be `build.tools`

Reply:
I'd try to keep it as explicit as possible if we can. That requires knowing how it will be used, which I don't fully understand. Is there a reason we're only defining this for `asdf` packages? Would an additional cache in the future not use the same mechanism? The most obvious answer to me is `build.executable_version.python` or similar. What is the exact class of thing that is defined here?  Is it different than `python.version`, or something else?

Reply:
Good feedback here!

@stsewd 

> could also be `build.tools`

I think this is the best option for now for me. Actually, `asdf` allows to install of things that are not "languages" but tools, so 👍🏼  on this.

@ericholscher 

> That requires knowing how it will be used, which I don't fully understand.

How I can help here? Is there a particular question you have? Maybe a quick call could work. In the meantime, I will answer the questions you have asked in the same comment.

> Is there a reason we're only defining this for `asdf` packages? Would an additional cache in the future not use the same mechanism?

I started thinking about "how to define the languages I want available in my build?", but from feedback it was risen that maybe language it's not a good idea, even if it's not implemented with `asdf` behind the scenes.

In fact, `asdf` support _other things_ that are not languages, but more like a tool. Example: `CMake`. This is the list of all plugins supported on `asdf` currently: https://giters.com/jthegedus/asdf-plugins

Because of that, I liked Santos' `build.tools` suggestion and I think we should use that one:

```yaml
build:
  tools:
    - python: 3.9
    - cmake: 3.12.2
    - nodejs: 14
```

> The most obvious answer to me is `build.executable_version_cache.python` or similar.

`.executable_` does not work to me because some packages could bring more than one executable (python -> python, pip, etc), or the name of "language" does not match the name of the "executable" (rust -> cargo)

Also, I don't think we have to expose the `.*_cache` part to users. Why it would be useful for them to know about it is cached?

> What is the exact class of thing that is defined here?

We are defining 2 things:

- what tools are going to be available in the builds
- what version of each tool

> Is it different than `python.version`, or something else?

It's not. `python.version` does the same work: "define what version of a pre-installed tool will be available". However, in the new Docker images, there won't be pre-installed tools; only the "runtime versions manager" (`asdf`) will be pre-install, which will be used to install the tools defined by the user.

@stsewd @ericholscher @agjohnson if we agree here, I will continue the work using `build.tools` everywhere and rename current things named as `build.languages`.

Reply:
Also, not sure where we can discuss this, but I think users will find confusing having the `python.version` not being used. And having two places for declaring the python env.

What about moving everything?

```yaml
version: 2
build:
  tools:
    python:
      version: 3.10
      install: ...
      system_packages: ...
``` 

Of course, people can use `python: 3.10` as a short form.

What I don't like is having one more nesting level, but I can live with that...

Reply:
Wow, i didn't realize asdf had so many plugins -- though not sure why I'd want multiple version of `direnv` installed :shrug: 

I like `build.tools`, it fits pretty well and has unique meaning.

I also just realized `build.addons` fits our use case pretty well too -- it's generic enough to cover languages and tools, lightly implies packaging, but doesn't overlap with system packages installed via `build.packages`.

I see what @stsewd is saying here about these being closer to "packages", as asdf is just a lightweight package manager. But it would seem hard to split up `build.packages` installable through apt and `build.packages` installable through asdf.

My RCV ballot would look like:

1) build.addons
2) build.tools
3) build.languages

Reply:
> Also, not sure where we can discuss this, but I think users will find confusing having the `python.version` not being used. And having two places for declaring the python env.
> What about moving everything?

I thought about this already and I didn't want to mention it yet because the change was not trivial and thought it may be good for the next iteration with a `version: 3` of the config file. For now, `python` and `conda` are treated as separated special cases. If we move them inside the new `build.tools.python` we will have different keys allowed depending the value of it.

Besides, they will still be special cases in comparison with `build.tools.nodejs` for example where you won't be able to specify dependencies to install.

Summarizing, I think it's a _big_ change that it's better to work later when `version: 3` for the Config File is implemented and does not block the progress we are doing here to move forward with new Docker images and languages/tools/addons.

Reply:
@agjohnson 

> I see what @stsewd is saying here about these being closer to "packages", as asdf is just a lightweight package manager

Well, they call themselves "Manage multiple runtime versions with a single CLI tool" (https://asdf-vm.com/). So, your original idea for `build.runtime` wasn't wrong --but I feel it's too technical maybe and not user-friendly.

> * build.addons
> * build.tools
> * build.languages

To me `addons` is interpreted as "plugins" or something pretty small that is attached to another program to make it behaves differently or to add an extra feature. I wouldn't say that "Python is an addon on Read the Docs builds", but I'd say that "Python is a tool added to Read the Docs builds".

So, `build.addons` is not _bad_ or _wrong_, but I just still prefer `build.tools` 😄

Reply:
I'm fine with `tools`, but it seems like we're basically trying to come up with a single name for 2 different things (eg. Python & Conda). `tools` at least gives us room to expand what we support there in the future, so I'm 👍  on it.

Reply:
Looks like `build.tools` is our top choice!

> I thought about this already and I didn't want to mention it yet

Hah yeah I was curious about this one as well but didn't want to complicate things yet.

Seems we could make the version 2 config file `python.version` also supply the value for `build.tools.python` if we wanted.

Would we also want to at least do some validation and test cases here, for when the user has a conflicting config file?

```yaml
build:
  tools:
    python: 3.9
python:
  version: 3.6
```

Reply:
> Would we also want to at least do some validation and test cases here, for when the user has a conflicting config file?

Yes, this example should produce an error. I wrote this in the document:

> python.version is incompatible with build.os or build.languages and will produce an error

(from https://docs.readthedocs.io/en/stable/development/design/build-images.html#specifying-extra-languages-requirements)

Reply:
> Looks like `build.tools` is our top choice!

Great!

I'm marking this conversation as resolved, to avoid showing all these comments while doing review.